### PR TITLE
Add sample for c++ client

### PR DESCRIPTION
**Pull request by Xiaohongt**

### CODE REVIEW

**Comment by msotclas** [comment]
Hi __@Xiaohongt__, I'm your friendly neighborhood Microsoft Open Technologies, Inc. Pull Request Bot (You can call me MSOTBOT). Thanks for your contribution!
    <span>You've already signed the contribution license agreement. Thanks!</span>
        <p>The agreement was validated by Microsoft Open Technologies, Inc. and real humans are currently evaluating your PR.</p>

TTYL, MSOTBOT;

**Comment by moozzyk** [comment]
Let's talk - I have quite comprehensive samples in this repo: https://github.com/moozzyk/SignalRClientCppSample which I am planning to use in a blog post introducing SignalR C++ Client. However in this repo we are missing some end to end tests and I think spending time on this might be more beneficial.

File: sample/sample/sample.cpp [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+#include "connection.h"
+#include "hub_connection.h"
+
+using namespace signalr;
+using namespace std;
+using namespace utility;
+
+void runRawConnection()
+{
+	connection * conn = new connection(L"http://localhost:40476/raw-connection");
```
Comment (moozzyk) [inline sample/sample/sample.cpp]: don't use raw pointers

File: sample/sample/sample.cpp [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+
+	conn->set_message_received([conn](const string_t& payload)
+	{
+		cout << "received message:" << endl;
+		wcout << payload << endl;
+
+		if (payload.find(L"stop") != string::npos)
+		{
+			cout << "stop connection" << endl;
+			conn->stop().wait();
```
Comment (moozzyk) [inline sample/sample/sample.cpp]: While this should work I think stopping connection from inside the message received callback is potentially risky. Also because the message received callback is invoked synchronously you should avoid blocking/spending to much time in this callback.

File: sample/sample/sample.cpp [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+		}
+	});
+
+	conn->start().wait();
```
Comment (moozzyk) [inline sample/sample/sample.cpp]: don't block. Instead use continuation (`.then`)

File: sample/sample/sample.cpp [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+
+	web::json::value obj;
+	obj[L"type"] = web::json::value::number(1);
+	obj[L"value"] = web::json::value::string(L"first message");
+
+	conn->send(obj.serialize()).wait();
```
Comment (moozzyk) [inline sample/sample/sample.cpp]: don't block - this probably can even be fire-and-forget

File: sample/sample/sample.cpp [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+
+	string str;
+
+	do
+	{
+		//send message
+		getline(cin, str);
+		obj[L"value"] = web::json::value::string(conversions::to_string_t(str));
+
+		if (conn->get_connection_state() == connection_state::connected)
```
Comment (moozzyk) [inline sample/sample/sample.cpp]: why is this needed?

File: sample/sample/sample.cpp [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+		{
+			conn->send(obj.serialize()).wait();
```
Comment (moozzyk) [inline sample/sample/sample.cpp]: same as above

File: sample/sample/sample.cpp (hunk 1 above)
Comment (moozzyk) [inline sample/sample/sample.cpp]: where is this server?

File: sample/sample/sample.vcxproj [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+    <LinkIncremental>false</LinkIncremental>
+  </PropertyGroup>
+  <ItemDefinitionGroup Condition="'$(Configuration)|$(Platform)'=='Debug|Win32'">
+    <ClCompile>
+      <PrecompiledHeader>Use</PrecompiledHeader>
+      <WarningLevel>Level3</WarningLevel>
+      <Optimization>Disabled</Optimization>
+      <PreprocessorDefinitions>WIN32;_DEBUG;_CONSOLE;_LIB;NO_SIGNALRCLIENT_EXPORTS;%(PreprocessorDefinitions)</PreprocessorDefinitions>
+      <SDLCheck>true</SDLCheck>
+      <AdditionalIncludeDirectories>..\..\src\signalrclient;..\..\include\signalrclient;..\..\packages\cpprestsdk.2.4.0.1\build\native\include;%(AdditionalIncludeDirectories)</AdditionalIncludeDirectories>
```
Comment (moozzyk) [inline sample/sample/sample.vcxproj]: The ways VS creative native projects is quite weird. Note that in this case you are only including AdditionalIncludeDirectories for `Win32|Debug` so as soon as someone change the Configuration to `Release` or the platform to `x64` the project won't compile. I always need to update the project manually to get it compile correctly regardless of the Confiugration or Platform

File: sample/sample/sample.cpp (hunk 3 above)
Comment (moozzyk) [inline sample/sample/sample.cpp]: In case of SignalR client it does not matter match but in general you should use `.get()` - AFAIR .wait() won't break on cancellations which, if not handled, will crash your process.

File: sample/sample/sample.cpp [hunk 8]
```diff
@@ -0,0 +1,86 @@
+// sample.cpp : Defines the entry point for the console application.
+//
+
```
Comment (moozzyk) [inline sample/sample/sample.cpp]: As opposed to C# in C++ async any time a task may throw (which in case of SignalR C++ Client is true any public method returning a task) you need to observe exceptions. If you don't the unobserved exception will crash your process.
